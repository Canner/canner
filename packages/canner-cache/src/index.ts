import {CacheI, Action, Listeners, Reducer} from './types';

export default class Cache implements CacheI {
  reducer: (data: any, action: Action) => any;
  data: Record<string, any>;
  listeners: Listeners;
  constructor({reducer}: {reducer: Reducer}) {
    this.reducer = (data, actions) => {
      // action array
      if (Array.isArray(actions)) {
        return actions.reduce(reducer, data);
      }
      // single action
      return reducer(data, actions);
    };
    this.data = {};
    this.listeners = {}
  }

  isCached = key => {
    // check the key is cached or not
    return Boolean(this.data[key]);
  };

  setData = (key, data) => {
    this.data[key] = data;
    this.publish(key);
  }

  getData = key => {
    if (this.isCached(key)) {
      return this.data[key];
    }
    throw new Error(`There is no cached data with key '${key}'`);
  };

  removeData = (key: string) => {
    delete this.data[key];
    delete this.listeners[key];
  }

  mutate = (key, actions) => {
    const originData = this.getData(key);
    const newData = this.reducer(originData, actions);
    this.setData(key, newData);
  }

  subscribe = (key, callback) => {
    const listener = {
      id: randomId(),
      callback
    };
    if (key in this.listeners) {
      this.listeners[key].push(listener) - 1;
    } else {
      this.listeners[key] = [listener];
    }
    return listener.id;
  }

  unsubscribe = (key, id) => {
    if (this.listeners[key]) {
      this.listeners[key] = this.listeners[key].filter(listener => listener.id !== id);
    }
  }

  publish = (key) => {
    const data = this.data[key];
    if (key in this.listeners) {
      this.listeners[key].forEach(listener => {
        listener.callback(data);
      });
    }
  }
}

function randomId(): string {
  return Math.random().toString(36).replace(/[^a-z]+/g, '').substr(2, 10);
}