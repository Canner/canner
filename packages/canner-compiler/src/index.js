import Traverser from './traverser';
import Parser from './parser';

export default {
  Traverser,
  Parser,
};
