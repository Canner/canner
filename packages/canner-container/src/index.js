export default, {transformSchemaToMenuConfig } from './components';
export { default as Sidebar } from './components/Sidebar';
export { default as Navbar } from './components/Navbar';
