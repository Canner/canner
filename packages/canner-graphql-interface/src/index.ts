import { createLink } from './link';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache, NormalizedCacheObject } from 'apollo-cache-inmemory';
import { Connector } from './connector/types';
import { CustomizeResolver } from './resolver';
import { GraphqlClient } from './graphqlClient/types';
import FirebaseRtdbClientConnector from './connector/firebaseConnector';
import FirestoreClientConnector from './connector/firestoreConnector';
import FirestoreAdminConnector from './connector/firestoreAdminConnector';
import MemoryConnector from './connector/memoryConnector';
import LocalStorageConnector from './connector/localStorageConnector';
import CannerPlatformConnector from './connector/cannerPlatformConnector';
import PrismaClient from './graphqlClient/prismaClient';
import GraphqlClientImpl from './graphqlClient/graphqlClient';

class FirebaseRtdbAdminConnector extends CannerPlatformConnector {
  constructor({namespace, projectId}: {namespace?: string; databaseURL: string; projectId: string}) {
    super();
    this.service = 'firebase';
    this.resourceId = projectId;
    this.connectorName = 'firebase.rtdb.admin';
    this.connectorParams = {
      namespace
    };
  }
}

const createClient = ({
  schema,
  connector,
  connectors,
  resolvers,
  graphqlClient,
}: {
  schema: any,
  connectors?: {[key: string]: Connector};
  resolvers?: {[key: string]: CustomizeResolver};
  graphqlClient?: GraphqlClient;
  connector?: Connector;
}) => {
  const apolloCache = new InMemoryCache();
  return new ApolloClient({
    cache: apolloCache,
    link: createLink({schema, connectors, graphqlClient, connector, resolvers})
  });
};

export {
  createClient,
  MemoryConnector,
  LocalStorageConnector,
  FirebaseRtdbAdminConnector,
  FirebaseRtdbClientConnector,
  FirestoreAdminConnector,
  FirestoreClientConnector,
  PrismaClient,
  GraphqlClientImpl as GraphqlClient
};
