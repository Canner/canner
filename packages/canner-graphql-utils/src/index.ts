import { filter } from './filter';
import paginator from './paginator';
import { Connector } from './types';

export {
  filter,
  paginator,
  Connector
};
