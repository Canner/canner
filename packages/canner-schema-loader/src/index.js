import {getOptions, stringifyRequest} from 'loader-utils';
import path from 'path';
import objectPath from 'object-path';
import Module from 'module';
import fs from 'fs';
import stripBOM from 'strip-bom';

const componentMap = {
  _map: {
    string: {
      input: '@canner/antd-string-input',
      card: '@canner/antd-string-card',
      link: '@canner/antd-string-link',
      radio: '@canner/antd-string-radio',
      select: '@canner/antd-string-select',
      textarea: '@canner/antd-string-textarea',
      time: '@canner/antd-string-time_picker',
      markdown: '@canner/antd-string-mde',
      default: '@canner/antd-string-input'
    },
    number: {
      input: '@canner/antd-number-input',
      rate: '@canner/antd-number-rate',
      slider: '@canner/antd-number-slider',
      default: '@canner/antd-number-input'
    },
    array: {
      tab: '@canner/antd-array-tabs',
      tabs: '@canner/antd-array-tabs',
      tabTop: '@canner/antd-array-tab_top',
      tabRight: '@canner/antd-array-tab_right',
      tabLeft: '@canner/antd-array-tab_left',
      tabBottom: '@canner/antd-array-tab_bottom',
      gallery: '@canner/antd-array-gallery',
      table: '@canner/antd-array-table',
      slider: '@canner/antd-array-slider',
      panel: '@canner/antd-array-panel',
      tag: '@canner/antd-array-tag',
      tableRoute: '@canner/antd-array-table_route',
      default: '@canner/antd-array-table'
    },
    object: {
      variants: '@canner/antd-object-variants',
      options: '@canner/antd-object-options',
      editor: '@canner/antd-object-editor',
      fieldset: '@canner/antd-object-fieldset',
      default: '@canner/antd-object-fieldset'
    },
    relation: {
      singleSelect: '@canner/antd-relation-single_select',
      multipleSelect: '@canner/antd-relation-multiple_select',
      default: '@canner/antd-relation-single_select'
    },
    boolean: {
      switch: '@canner/antd-boolean-switch',
      card: '@canner/antd-boolean-card',
      default: '@canner/antd-boolean-switch'
    },
    geoPoint: {
      default: '@canner/antd-object-map'
    },
    file: {
      file: '@canner/antd-object-image',
      default: '@canner/antd-object-image'
    },
    image: {
      image: '@canner/antd-object-image',
      default: '@canner/antd-object-image'
    },
    dateTime: {
      dateTime: '@canner/antd-string-date_time_picker', 
      default: '@canner/antd-string-date_time_picker'
    }
  },
  get: function(type, ui = 'default') {
    if (!(type in this._map)) {
      return undefined;
    }

    if (!(ui in this._map[type])) {
      throw new Error(`there is no ui ${ui} in type ${type}`);
    }

    return objectPath.get(this._map, `${type}.${ui}`, ui);
  },
  set: function(path, name) {
    const paths = path.split('.');
    objectPath.set(this._map, paths, name);
  }
};
function nullReplace(match, type) {
  const packageName = componentMap.get(type);
  return match.replace('null', `{packageName: '${packageName}'}`);
}
function getPackageName(match, type, attrs) {
  const result = /ui: ['"](\w+)['"]/.exec(match);
  let packageName = '';
  if (result === null) {
    packageName = componentMap.get(type);
  } else {
    packageName = componentMap.get(type, result[1]);
  }
  if (!packageName) {
    return match;
  }
  if (attrs === 'null') {
    return match.replace(/['"](string|number|boolean|array|object|relation|geoPoint|dateTime|image)['"], null/g, nullReplace);
  } else {
    var braceIndex = match.indexOf('{');
    if (braceIndex === -1) {
      return match;
    }
    return `${match.substring(0, braceIndex + 1)}packageName: '${packageName}', ${match.substring(braceIndex + 1)}`;
  }
}

function packageNameTemplate(match, source, context) {
  /** fieldset is not a real component, so return its name, and Generator will deal with it */
  if (!source || source.endsWith('@canner/antd-object-fieldset')) {
    return match;
  }
  let packageName = source;
  if (packageName.startsWith('.') || packageName.startsWith('/')) {
    // customize component
    packageName = path.resolve(context.context, source);
  } else {
    // deployed component
    const paths = require.resolve(packageName).split(path.sep);
    const dirPathIndex = paths.indexOf(packageName);
    packageName = paths.slice(0, dirPathIndex - 1).join('/');
  }
  const loader = ['new Promise((resolve) => {',
    'require.ensure([], (require) => {',
      `resolve(require("${packageName}"));`,
    '});',
  '})'].join('');
  let builder;
  const defFile = `${packageName}/canner.def.js`;
  if (fs.existsSync(defFile)) {
    // to check defFile is exist or not, if not this line will throw an error
    builder = `require("${defFile}").default ? require("${defFile}").default : require("${defFile}")`;
  } else {
    builder = undefined;
  }

  let canner = {};
  try {
    canner = require(`${packageName}/package.json`).canner || {};
  } catch (e) {
    canner = {};
  }
  canner = Object.keys(canner).reduce((result, key, currentIndex, arr) => {
    const last = arr.length - 1 === currentIndex;
    if (key === 'cannerDataType') {
      result += `type: '${canner[key]}'${last ? '' : ','}`;
    } else {
      result += `${key}: ${typeof canner[key] === 'string' ? `'${canner[key]}'` : canner[key]}${last ? '' : ','}`;
    }
    return result;
  }, '');
  return [`packageName: '${packageName}',`,
    `loader: ${loader}`,
    `${builder ? `, builder: ${builder}` : ''}${canner ? `, ${canner}`: ''}`
  ].join('');
}

export default function loader(source) {
  const options = getOptions(this);
  const cannerComponentMap = {};
  // get type and ui and add packagename in code
  let str = replaceTypeAndUItoPackageName(source);
  // transform packageName to loader, builder, and other attributes in package.json
  str = replacePackageNameToDynamicImport(str, this);
  return str;
}

function replacer(key, val) {
  if (key === '__CANNER_KEY__') {
    return [];
  }

  return val;
}

export function replaceTypeAndUItoPackageName(str) {
  return str.replace(/\s*?['"](string|number|boolean|array|object|relation|geoPoint|dateTime|image)['"],\s*?(({\s*?((?!(packageName)).|\s)*?})|null)/g, getPackageName);
}

export function replacePackageNameToDynamicImport(str, that) {
  return str.replace(/packageName: ['"]([\w\/\-\@\.]+)['"]/g, (match, name) => packageNameTemplate(match, name, that))
}
