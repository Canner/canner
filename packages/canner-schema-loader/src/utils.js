import objectPath from 'object-path';
export const componentMap = {
  _map: {
    string: {
      input: '@canner/antd-string-input',
      card: '@canner/antd-string-card',
      link: '@canner/antd-string-link',
      radio: '@canner/antd-string-radio',
      select: '@canner/antd-string-select',
      textarea: '@canner/antd-string-textarea',
      time: '@canner/antd-string-time_picker',
      markdown: '@canner/antd-string-mde',
      default: '@canner/antd-string-input'
    },
    number: {
      input: '@canner/antd-number-input',
      rate: '@canner/antd-number-rate',
      slider: '@canner/antd-number-slider',
      default: '@canner/antd-number-input'
    },
    array: {
      tab: '@canner/antd-array-tabs',
      tabs: '@canner/antd-array-tabs',
      tabTop: '@canner/antd-array-tab_top',
      tabRight: '@canner/antd-array-tab_right',
      tabLeft: '@canner/antd-array-tab_left',
      tabBottom: '@canner/antd-array-tab_bottom',
      gallery: '@canner/antd-array-gallery',
      table: '@canner/antd-array-table',
      slider: '@canner/antd-array-slider',
      panel: '@canner/antd-array-panel',
      tag: '@canner/antd-array-tag',
      tableRoute: '@canner/antd-array-table_route',
      tree: '@canner/antd-array-tree',
      default: '@canner/antd-array-table'
    },
    object: {
      variants: '@canner/antd-object-variants',
      options: '@canner/antd-object-options',
      editor: '@canner/antd-object-editor',
      fieldset: '', // no package
      default: '' // no package
    },
    relation: {
      singleSelect: '@canner/antd-relation-single_select',
      multipleSelect: '@canner/antd-relation-multiple_select',
      singleSelectTree: '@canner/antd-relation-single_select_tree',
      default: '@canner/antd-relation-single_select'
    },
    boolean: {
      switch: '@canner/antd-boolean-switch',
      card: '@canner/antd-boolean-card',
      default: '@canner/antd-boolean-switch'
    },
    geoPoint: {
      default: '@canner/antd-object-map'
    },
    file: {
      file: '@canner/antd-object-image',
      default: '@canner/antd-object-image'
    },
    image: {
      image: '@canner/antd-object-image',
      default: '@canner/antd-object-image'
    },
    dateTime: {
      dateTime: '@canner/antd-string-date_time_picker', 
      default: '@canner/antd-string-date_time_picker'
    },
    chart: {
      line: '@canner/vega-chart-line',
      bar: '@canner/vega-chart-bar',
      pie: '@canner/vega-chart-pie',
      scatter: '@canner/vega-chart-scatter',
      default: '@canner/vega-chart-line'
    },
    indicator: {
      list: '@canner/antd-indicator-list',
      amount: '@canner/antd-indicator-amount',
      default: '@canner/antd-indicator-amount'
    },
    enum: {
      select: '@canner/antd-string-select',
      default: '@canner/antd-string-select'
    }
  },
  get: function(type, ui = 'default') {
    if (!(type in this._map)) {
      return undefined;
    }

    if (!(ui in this._map[type])) {
      throw new Error(`there is no ui ${ui} in type ${type}`);
    }

    return objectPath.get(this._map, `${type}.${ui}`, ui);
  },
  set: function(path, name) {
    const paths = path.split('.');
    objectPath.set(this._map, paths, name);
  }
};