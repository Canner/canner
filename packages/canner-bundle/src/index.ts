import webpack from 'webpack';
import WebpackDevServer from 'webpack-dev-server';
import {createConfig, CreateConfigArgsType, createWebConfig, createSchemaConfig} from './utils/createWebpackConfig';
import transformSchemaToJson from './utils/transformSchemaToJSON';
import {SCHEMA_OUTPUT_PATH, SCHEMA_JSON_OUTPUT_PATH} from './config';
export function build(options?: CreateConfigArgsType) {
  const config = createConfig(options || {});
  return new Promise((resolve, reject) => {
    webpack(config)
      .run((err: any, stats) => {
        if (err) {
          console.error(err.stack || err);
          if (err.details) {
            return reject(err.details);
          }
        }
      
        const info = stats.toJson();
      
        if (stats.hasErrors()) {
          return reject(info.errors);
        }
        if (!options.webOnly) {
          const schemaPath = `${(options || {}).schemaOutputPath || SCHEMA_OUTPUT_PATH}/schema.node.js`;
          transformSchemaToJson(schemaPath, (options || {}).schemaJsonOutputPath || SCHEMA_JSON_OUTPUT_PATH);
        }
        resolve(stats);
      });
  });
}

export function watchSchema(options: CreateConfigArgsType, callback: any) {
  if (options && !options.watch) {
    options.watch = true;
  }
  const config = createSchemaConfig(options || {});
  const watchOptions = typeof config.watch === 'object'?
    config.watch :
    {
      aggregateTimeout: 1000,
    };
  return webpack(config)
    .watch(watchOptions, (err, stats) => {
      try {
        const schemaPath = `${(options || {}).schemaOutputPath || SCHEMA_OUTPUT_PATH}/schema.node.js`;
        transformSchemaToJson(schemaPath, (options || {}).schemaJsonOutputPath || SCHEMA_JSON_OUTPUT_PATH);
      } catch(e) {
        err = e;
      }
      callback(err, stats);
    });
}

export function createWebpackDevServer(options: CreateConfigArgsType) {
  if (options) {
    options.watch = true;
  }
  const config = createWebConfig(options || {}) as any;
  config.entry.index.unshift(`webpack-dev-server/client?http://localhost:${config.devServer.port}/`, `webpack/hot/dev-server`);
  const compiler = webpack(config);
  const devServerOptions = {...config.devServer};
  WebpackDevServer.addDevServerEntrypoints(config, devServerOptions);
  const server = new WebpackDevServer(compiler, devServerOptions);
  return {
    compiler,
    server,
    config
  };
}
