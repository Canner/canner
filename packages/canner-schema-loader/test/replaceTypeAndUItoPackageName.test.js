import {replaceTypeAndUItoPackageName} from '../src';

describe('type and ui to packageName', () => {
  it('should get default packageName', () => {
    const str = `
      builder('string', null);
    ` 
    expect(replaceTypeAndUItoPackageName(str)).toMatchSnapshot();
  });

  it('should get default packageName', () => {
    const str = `
      builder('string', {
        keyName: 'string'
      });
    ` 
    expect(replaceTypeAndUItoPackageName(str)).toMatchSnapshot();
  });

  it('should get select packageName', () => {
    const str = `
      builder('string', {
        keyName: 'string',
        ui: 'select'
      });
    ` 
    expect(replaceTypeAndUItoPackageName(str)).toMatchSnapshot();
  });

  it('should not change', () => {
    const str = `
      builder('string', {
        keyName: 'string',
        packageName: '/packages/select'
      });
    ` 
    expect(replaceTypeAndUItoPackageName(str)).toBe(str);
  });

  it('should get package name', () => {
    const str = `
      builder('object', {keyName: 'info'},
        builder('string', {
          keyName: 'string'
        });
      );
    ` 
    expect(replaceTypeAndUItoPackageName(str)).toMatchSnapshot();
  });

  it('should get package name', () => {
    const str = `
      builder('object', {keyName: 'info'},
        builder('string', null);
      );
    ` 
    expect(replaceTypeAndUItoPackageName(str)).toMatchSnapshot();
  });
});