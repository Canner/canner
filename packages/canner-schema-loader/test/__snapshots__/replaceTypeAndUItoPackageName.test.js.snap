// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type and ui to packageName should get default packageName 1`] = `
"
      builder('string', {packageName: '@canner/antd-string-input'});
    "
`;

exports[`type and ui to packageName should get default packageName 2`] = `
"
      builder('string', {packageName: '@canner/antd-string-input', 
        keyName: 'string'
      });
    "
`;

exports[`type and ui to packageName should get package name 1`] = `
"
      builder('object', {packageName: '@canner/antd-object-fieldset', keyName: 'info'},
        builder('string', {packageName: '@canner/antd-string-input', 
          keyName: 'string'
        });
      );
    "
`;

exports[`type and ui to packageName should get package name 2`] = `
"
      builder('object', {packageName: '@canner/antd-object-fieldset', keyName: 'info'},
        builder('string', {packageName: '@canner/antd-string-input'});
      );
    "
`;

exports[`type and ui to packageName should get select packageName 1`] = `
"
      builder('string', {packageName: '@canner/antd-string-select', 
        keyName: 'string',
        ui: 'select'
      });
    "
`;
